### DISCUSSION TITLE
How to globally handle key events in Avalonia 11?

### DISCUSSION THREAD

**Discussion by gusmanb** [AvaloniaUI/Avalonia#12179] [Q&A] [answered]
Hi.

Previous to the new Avalonia release I could use the InputManager to trap raw input events like key events and text input events.
In Avalonia 11 the InputManager class is now internal and cannot be used to do this.

Let me explain my situation. I have a single view with multiple subviews, the program is an IDE for zx spectrum software development. When the IDE starts running the integrated emulator I trap any key/text event and route it to the emulator so the user does not need to focus it.

I have tried to attach to the KeyUp/KeyDown/TextInput events but as it receives only those through bubbling there are some controls that intercept the event and it is not received by the window.

Also, something that makes the case even worse is that I have created a docking system for the IDE that allows me to make controls to float, so they are moved to a different window and the events need to be redirected to the emulator even if they come from a different window or the emulator is floated...

How can I intercept those events now that the InputManager is not available?

Cheers.

**Comment by danielmayost** [comment]
i think you need to subscribe to those event on `TopLevel`, for example:
```
var topLevel = TopLevel.GetTopLevel(this)!;
topLevel.KeyDown += OnKeyDown;
```

**Reply by gusmanb** [reply]
I am subscribed on the toplevel, but the toplevel receives the messages through bubbling, so any control that flags the events as handled blocks the toplevel from receiving them. Also, I create dynamically floating windows with the dock system so even if that worked I would need to subscribe to all the events each time a new window is created...

**Reply by maxkatz6** [reply]
> but the toplevel receives the messages through bubbling

Subscribe then with Tunnel strategy. You can specify in AddEventHandler method.

**Reply by danielmayost** [reply, marked as answer]
**Edit:** @maxkatz6  Sorry, I didn't see your comment when I posted.

IIRC, you can workaround bubbling by:

```
InputElement.KeyDownEvent.AddClassHandler<TopLevel>(OnKeyDown, handledEventsToo: true);
```

As for listing per window, how would you do it in InputManager before?

**Reply by gusmanb** [reply]
@maxkatz6 I tried it, but not sure why the events fired all after being processed, I mean, when I press a key, hold it pressed and then after one or two seconds I released it the keyup/down events were not received until I released the key and then both were received at the same time, very strange.

@danielmayost Hmmm, that will register the event for all the TopLevel instances and receive even handled events? If it does then that might be the solution, I'm going to try it :D
Before, with the InputManager I had an `IObserver<RawInputEventArgs>` and that received every event in the application, it is not associated to any control but to the raw input.

**Reply by gusmanb** [reply]
@danielmayost It works like a charm! Thanks a lot!
